core/src/storage/lazy_chunk.rs

```diff
@@ -0,0 +1,393 @@
+    /// Takes the value from the entry and returns it.
+    ///
+    /// # Note
+    ///
+    /// This changes the `mutate` state of the entry if the entry was occupied.
+    pub fn take_value(&mut self) -> Option<T> {
+        self.mutated.set(self.value.is_some());
+        self.value.take()
+    }
+
+    /// Puts the new value into the entry and returns the old value.
+    ///
+    /// # Note
+    ///
+    /// This changes the `mutate` state of the entry to `true` as long as at
+    /// least one of `old_value` and `new_value` is `Some`.
+    pub fn put(&mut self, new_value: Option<T>) -> Option<T> {
+        match new_value {
+            Some(new_value) => {
+                self.mutated.set(true);
+                self.value.replace(new_value)
+            }
+            None => self.take_value(),
+        }
+    }
+}
+
+impl<T> LazyChunk<T> {
+    /// Creates a new empty lazy chunk that cannot be mutated.
+    pub fn new() -> Self {
+        Self {
+            key: None,
+            cached_entries: UnsafeCell::new(EntryMap::new()),
+        }
+    }
+
+    /// Performs the given closure on the mutable cached entries.
+    ///
+    /// # Note
+    ///
+    /// Actions on the mutable lazy entries are performed within the closure
+    /// to not leak exclusive references to it to the outside. This is important
+    /// since the `for_entries` method itself operates only on `&self`.
+    fn for_cached_entries<F, R>(&self, f: F) -> R
+    where
+        F: FnOnce(&mut EntryMap<T>) -> R,
+    {
+        // SAFETY: We operate on an exclusive reference on `BTreeMap` within the
+        //         given closure while our method receiver is only a shared reference.
+        //         However, due to encapsulation of the exclusive reference within
+        //         the given closure we cannot leak the exclusive reference outside
+        //         of the closure. So the below action is safe in this regard.
+        f(unsafe { &mut *self.cached_entries.get() })
+    }
+}
+
+impl<T> StorageSize for LazyChunk<T> {
+    const SIZE: u64 = <Chunk as StorageSize>::SIZE;
+}
+
+impl<T> Pull for LazyChunk<T>
+where
+    Self: StorageSize,
+{
+    fn pull(key_ptr: &mut KeyPtr) -> Self {
+        Self {
+            key: Some(key_ptr.next_for::<Self>()),
+            cached_entries: UnsafeCell::new(BTreeMap::new()),
+        }
+    }
+}
+
+impl<T> Push for LazyChunk<T>
+where
+    T: Unpin + Push + StorageSize + scale::Encode,
+{
+    fn push(&self, key_ptr: &mut KeyPtr) {
+        let key = key_ptr.next_for::<Self>();
+        assert_eq!(self.key, Some(key));
+        self.for_cached_entries(|entries| {
+            for (&index, entry) in entries.iter_mut().filter(|(_, entry)| entry.mutated())
+            {
+                let offset: Key = key + index;
+                let mut ptr = KeyPtr::from(offset);
+                Push::push(&**entry, &mut ptr);
+            }
+        });
+    }
+}
+
+impl<T> LazyChunk<T>
+where
+    T: Unpin + scale::Decode,
```

Comment:
You don't need `Unpin` because you never prove to `T` that you have pinned it. You never create a `Pin<&mut T>` or `Pin<&T>` or anything similar anywhere.

Reply:
@tomaka Actually Rust compiler thinks that I need it:

```
error[E0596]: cannot borrow data in a dereference of `core::pin::Pin<env::types::_::_parity_scale_codec::alloc::boxed::Box<storage::lazy_chunk::Entry<T>>>` as mutable
   --> core/src/storage/lazy_chunk.rs:248:50
    |
248 |             BTreeMapEntry::Occupied(occupied) => &mut **occupied.into_mut(),
    |                                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot borrow as mutable
    |
    = help: trait `DerefMut` is required to modify through a dereference, but it is not implemented for `core::pin::Pin<env::types::_::_parity_scale_codec::alloc::boxed::Box<storage::lazy_chunk::Entry<T>>>`

error[E0596]: cannot borrow data in a dereference of `core::pin::Pin<env::types::_::_parity_scale_codec::alloc::boxed::Box<storage::lazy_chunk::Entry<T>>>` as mutable
   --> core/src/storage/lazy_chunk.rs:256:17
    |
256 | /                 &mut **vacant.insert(Box::pin(Entry {
257 | |                     value,
258 | |                     mutated: core::cell::Cell::new(false),
259 | |                 }))
    | |___________________^ cannot borrow as mutable
    |
    = help: trait `DerefMut` is required to modify through a dereference, but it is not implemented for `core::pin::Pin<env::types::_::_parity_scale_codec::alloc::boxed::Box<storage::lazy_chunk::Entry<T>>>`
```

Reply:
`impl<T> Unpin for Entry<T> {}`

Or just don't use `Pin<Box<>>` but just `Box<>`.